### PR TITLE
Various additions to the standard libraries

### PR DESCRIPTION
**Pull request by truecolour**
This PR adds various constants, functions and macros that I've found useful while doing some toy projects in pixie, and slightly tweaks the behaviour of various others.

I tried to keep the commits mostly independent, so it should be easy to remove stuff if so desired.

I don't know much about PyPy; someone should probably take a closer look at the RPython parts (665e78b and 654829b) before merging them.

### CODE REVIEW

File: pixie/stdlib.pxi [hunk 1]
```diff
@@ -1568,8 +1589,21 @@ not enough elements were present."
...
-       (cons (take n s) (partition n step (drop step s))))))
+   (when-let [s (seq coll)]
+     (lazy-seq
+       (cons (take n s) (partition n step (drop step s)))))))
+
+(defn partitionf [f coll]
+  {:doc "A generalized version of partition. Instead of taking a constant number of elements,
+         this function calls f with the remaining collection to determine how many elements to
+         take."
+   :examples [["(fpartition first [1 :a, 2 :a b, 3 :a :b :c])"
```
Comment (halgari) [inline pixie/stdlib.pxi]: `fpartition` vs `partitionf` but I like where you're going with this

File: pixie/vm/stdlib.py [hunk 2]
```diff
@@ -337,16 +337,36 @@ def apply__args(args):
 @returns(bool)
 @as_var("instance?")
 def _instance(c, o):
-    affirm(isinstance(c, Type), u"c must be a type")
+    from pixie.vm.persistent_vector import PersistentVector
+    if isinstance(c, Type):
+        return true if istypeinstance(o, c) else false
+    elif isinstance(c, PersistentVector):
```
Comment (halgari) [inline pixie/vm/stdlib.py]: I'd prefer this be a check that something satisfies both ICounted and IIndexed. That way it supports things like arrays as well.

Comment (halgari) [inline pixie/vm/stdlib.py]: But +1 to the idea here

File: pixie/vm/stdlib.py [hunk 3]
```diff
@@ -337,16 +337,36 @@ def apply__args(args):
...
 @as_var("satisfies?")
 def _satisfies(proto, o):
-    affirm(isinstance(proto, Protocol), u"proto must be a Protocol")
-
-    return true if proto.satisfies(o.type()) else false
+    from pixie.vm.persistent_vector import PersistentVector
+    t = o.type()
+    if isinstance(proto, Protocol):
+        return true if proto.satisfies(t) else false
+    elif isinstance(proto, PersistentVector):
```
Comment (halgari) [inline pixie/vm/stdlib.py]: Same as with `instance?`

**Comment by halgari** [comment]
Pretty epic PR. Address the few comments I posted and I'll merge it.

File: pixie/vm/stdlib.py (hunk 2 above)
Comment (heyLu) [inline pixie/vm/stdlib.py]: Maybe we should introduce an internal `-instance?` (and `-satisfies?`) and then define this version in `stdlib.pxi`?

**Comment by truecolour** [comment]
@heyLu That was my idea as well.

I'll update the PR this evening.

**Comment by truecolour** [comment]
Actually, now that I think about it, the more common case for `satisfies?` is probably checking if a value satisfies all protocols, not any one of them. If there are no objections I'll include that change.

**Comment by heyLu** [comment]
A note regarding `condp` and `case`: I think they should throw if no clause matches and there is no default expression, as they do in Clojure. If someone wants to return nil they can provide it as the default expression.

**Comment by truecolour** [comment]
Here are the changes I promised.

I've got to say, working with a Lisp again is loads of fun ;-).

**Comment by halgari** [comment]
Thanks!

**Pull request merged**